src/ucp/rndv/proto_rndv.inl

```diff
@@ -234,32 +234,41 @@ ucp_proto_rndv_bulk_max_payload(ucp_request_t *req,
     size_t lane_offset, max_payload, scaled_length;
 
     /* Request remaining length cannot be lower than minimal payload size */
-    ucs_assertv((req_length - req_offset) >= min_frag,
-                "req=%p length-offset=(%zu-%zu)=%zu min_frag=%zu", req,
-                req_length, req_offset, req_length - req_offset, min_frag);
+    ucs_assertv(req_remainder >= min_frag,
+                "req=%p req_remainder=(%zu-%zu)=%zu min_frag=%zu", req,
+                req_length, req_offset, req_remainder, min_frag);
 
     if (ucs_likely(total_length < max_frag_sum)) {
         /* Each lane sends less than its maximal fragment size */
         scaled_length = ucp_proto_multi_scaled_length(lpriv->weight_sum,
                                                       total_length);
-        max_payload   = scaled_length - total_offset;
+
+        ucs_assertv(scaled_length >= total_offset,
+                    "req=%p scaled_length=%zu total_offset=%zu "
+                    "total_length=%zu weight_sum=%zu%% ",
+                    req, scaled_length, total_offset, total_length,
+                    ucp_proto_multi_scaled_length(lpriv->weight_sum, 100));
+
+        max_payload = scaled_length - total_offset;
     } else {
         /* Send in round-robin fashion, each lanes sends its maximal size */
         lane_offset = total_offset % max_frag_sum;
-
-        ucs_assertv(lpriv->max_frag_sum >= lane_offset,
-                    "req=%p lpriv->max_frag_sum=%zu lane_offset=%zu "
-                    "total_offset=%zu max_frag_sum=%zu ",
-                    req, lpriv->max_frag_sum, lane_offset, total_offset,
-                    max_frag_sum);
-
-        max_payload = lpriv->max_frag_sum - lane_offset;
+        if (ucs_likely(lane_offset < lpriv->max_frag_sum)) {
+            max_payload = lpriv->max_frag_sum - lane_offset;
+        } else {
+            /* The offset is before the start of this lane, it can happen if we
+               adjusted the previous lane max_payload to accomodate min_frag */
+            max_payload = lpriv->max_frag;
+        }
     }
 
-    /* If there is some remaining data - ensure a minimal size of the next
-       fragment, by sending less data in the current fragment */
+    /* If there is some remaining data for the next lane to send - make sure we
+       do not send more than (remainder - min_frag) on the current lane.
+       This way, the next lane will always have at least min_frag bytes to send.
+     */
     if (max_payload < req_remainder) {
         max_payload = ucs_min(max_payload, req_remainder - min_frag);
+        max_payload = ucs_max(max_payload, min_frag);
```

Comment:
can `req_reminder` be less than `2*min_frag` here (maybe if `max_payload` is close to `min_frag`)? Then next lane will have less than `min_frag` to send

Reply:
reimplemented in a different way (also with less branches) - if we're going to send too small data, extend the send length (but do not change global request offset) to min_frag. this can potentially send a little more data than needed but is more efficient in cases there is no problem, and keeps the offset in a consistent state.